File: software/experimental/wxPython/newkey.py
import wx
import re

from util import Rect, Point, say, Size
from notifications import KeyUpdateNotification, ModelChangeNotification, \
                          LayerChangeNotification
from keydisplay import KeyDisplayPane, DisplayKey, DISPLAY_MODE_DYNAMIC_RESIZE, \
                       DISPLAY_MODE_WRAPPED
from model import Model, SourceKey, MacroKey
from hid import Usage


class NewKeyDropTarget(wx.DropTarget):

    def __init__(self, parent):
        wx.DropTarget.__init__(self)
        self.parent = parent
        self.data = wx.TextDataObject()
        self.SetDataObject(self.data)

#   def OnDrop(self, x, y):
#       dlg = wx.MessageDialog(self.parent, "Dropped at (%d, %d)" % (x, y),
#                              "Dropping", wx.OK | wx.ICON_INFORMATION)
#       dlg.ShowModal()
#       dlg.Destroy()

    def OnData(self, x, y, default):
        self.GetData()
        actualData = self.data.GetText()
        self.parent.DoDrop(x, y, actualData)
        return default


class TestPane(wx.Panel):

    def __init__(self, parent):
        wx.Panel.__init__(self, parent, size=(200, 200))
        self.Bind(wx.EVT_PAINT, self.OnPaint)

    def OnPaint(self, event):
        dc = wx.PaintDC(self)
        rect = self.GetClientRect()
        dc.SetPen(wx.Pen('black'))
        dc.DrawRectangleRect(rect)
        dc.DrawLabel("%s (%s, %s)" % (self.__class__.__name__, rect.width, rect.height), rect.Get(),
                     wx.ALIGN_CENTER)


class NewKeyPane(wx.Panel):

    def __init__(self, parent, keyHandle=None):
        wx.Panel.__init__(self, parent)

        self.okButton = wx.Button(self, wx.ID_OK)
        self.cancelButton = wx.Button(self, wx.ID_CANCEL)

        sourceKey = None
        if keyHandle is not None:
            sourceKey = Model.sourcekeys[keyHandle]
        else:
            sourceKey = SourceKey(None)
        assert(sourceKey is not None)

        dispPane  = DispPane(self, sourceKey)
        macroPane = MacroPane(self, sourceKey)
        modPane   = ModifierKeyPane(self)

        leftSizer = wx.BoxSizer(wx.VERTICAL)
        leftSizer.Add(dispPane, proportion=1, flag=wx.EXPAND)
        leftSizer.Add(modPane,  proportion=1, flag=wx.EXPAND)

        upperSizer = wx.BoxSizer(wx.HORIZONTAL)
        upperSizer.Add(leftSizer, proportion=1, flag=wx.EXPAND)
        upperSizer.Add(macroPane,  proportion=2, flag=wx.EXPAND)

        btnSizer = wx.StdDialogButtonSizer()
        btnSizer.AffirmativeButton = self.okButton
        btnSizer.CancelButton = self.cancelButton
        btnSizer.Realize()

        mainSizer = wx.BoxSizer(wx.VERTICAL)
        mainSizer.Add(upperSizer, proportion=1, flag=wx.EXPAND)
        mainSizer.Add(btnSizer, flag=wx.EXPAND|wx.ALL, border=10)
        self.SetSizer(mainSizer)

        mainSizer.Fit(self)
        mainSizer.SetSizeHints(self)


class DispPane(KeyDisplayPane):

    def __init__(self, parent, sourceKey):
        KeyDisplayPane.__init__(self, parent, border=10, displayMode=DISPLAY_MODE_DYNAMIC_RESIZE)
        self._sourceKey = sourceKey
        self._dispKey = DisplayKey(sourceKey.handle, Rect(0, 0, 1.8, 1.8))
        self._displayKeys = {sourceKey.handle:self._dispKey}
        self._updateKeys()
        KeyDisplayPane._initKeys(self)

    def _getContentSize(self):
        return Size(1.8, 1.8)

#   def OnPaint(self, event):
#       dc = wx.PaintDC(self)
#       rect = self.GetClientRect()
#       dc.DrawRectangleRect(rect)
#       KeyDisplayPane.OnPaint(self, event)

class MacroPane(KeyDisplayPane):

    def __init__(self, parent, sourceKey):
        KeyDisplayPane.__init__(self, parent, border=10, displayMode=DISPLAY_MODE_WRAPPED)
        self._displayKeys = {}
        KeyDisplayPane._initKeys(self)
        self.Bind(wx.EVT_PAINT, self.OnPaint)
        self._keySize = Size(6.0, 2.5)
        self._keySpacing = Size(0.2, 0.2)
        keyRect = Rect(0, 0, *self._keySize)
        handle = None
        if sourceKey.usage:
            handle = sourceKey.usage.MakeHandle()
        elif sourceKey.mode:
            handle = "mode:%s" % sourceKey.mode
        elif len(sourceKey.macro) > 0:
            for (i, macroKey) in enumerate(sourceKey.macro):
                handle = macroKey.MakeHandle(i)
                self._displayKeys[handle] = DisplayKey(handle, keyRect)
                self._orderedKeyHandles.append(handle)
            handle = None
        if handle is not None:
            dispKey = DisplayKey(handle, keyRect)
            self._displayKeys[handle] = dispKey
            self._orderedKeyHandles.append(handle)
            self._updateKeys()

    def _getLabels(self, displayKey):
        labels = {}
        print displayKey.handle
        keyInfo = re.split(':', displayKey.handle)
        if keyInfo[0] == 'usage':
            labels['TopLeft'] = "page: %s id: %s" % (keyInfo[1], keyInfo[2])
        elif keyInfo[0] == 'mode':
            labels['TopLeft'] ="mode: %s" % keyInfo[1]
        elif keyInfo[0] == 'macro':
            labels['BottomLeft'] = "modifiers: %s " % keyInfo[4]
            labels['TopLeft'] = "page: %s id: %s" % (keyInfo[2], keyInfo[3])
        elif keyInfo[0] == 'm_usage':
            labels['TopLeft'] = "page: %s id: %s" % (keyInfo[2], keyInfo[3])
        else:
            labels['BottomLeft'] = "?"
        return labels

    def OnPaint(self, event):
        dc = wx.PaintDC(self)
        rect = self.GetClientRect()
        dc.DrawRectangleRect(rect)
        KeyDisplayPane.OnPaint(self, event)


class ModifierKeyPane(KeyDisplayPane):

    def __init__(self, parent):
        KeyDisplayPane.__init__(self, parent, border=0, displayMode=DISPLAY_MODE_DYNAMIC_RESIZE)
        self._initKeys()

    def _initKeys(self):
        KeyDisplayPane._initKeys(self)
        self._displayKeys = { 'left_shift':  DisplayKey('left_shift',  Rect(0.0, 0.0, 2.8, 1.8)),
                              'left_ctrl':   DisplayKey('left_ctrl',   Rect(0.0, 2.0, 2.8, 1.8)),
                              'left_alt':    DisplayKey('left_alt',    Rect(0.0, 4.0, 2.8, 1.8)),
                              'left_gui':    DisplayKey('left_gui',    Rect(0.0, 6.0, 2.8, 1.8)),
                              'right_shift': DisplayKey('right_shift', Rect(3.0, 0.0, 2.8, 1.8)),
                              'right_ctrl':  DisplayKey('right_ctrl',  Rect(3.0, 2.0, 2.8, 1.8)),
                              'right_alt':   DisplayKey('right_alt',   Rect(3.0, 4.0, 2.8, 1.8)),
                              'right_gui':   DisplayKey('right_gui',   Rect(3.0, 6.0, 2.8, 1.8)) }
        self._updateKeys()

    def _getContentSize(self):
        return Size(6.0, 8.0)

def main():
    app = wx.PySimpleApp()
    frame = wx.Frame(None, -1, 'dest')
    pane = NewKeyPane(frame)
    frame.Show()
    app.MainLoop()

if __name__ == '__main__':

    main()

File: software/experimental/wxRuby/newkey.rb
#import wx
#import re
#
#from util import Rect, Point, say, Size
#from notifications import KeyUpdateNotification, ModelChangeNotification, \
#                          LayerChangeNotification
#from keydisplay import KeyDisplayPane, DisplayKey, DISPLAY_MODE_DYNAMIC_RESIZE, \
#                       DISPLAY_MODE_WRAPPED
#from model import Model, SourceKey, MacroKey
#from hid import Usage

require 'wx'
require 'keydisplay'

class NewKeyDropTarget < Wx::DropTarget

    def initialize(parent)
        super
        @parent = parent
        @data = Wx::TextDataObject()
        SetDataObject(@data)
    end

    def OnData(x, y, default)
        super
        actualData = @data.GetText()
        @parent.DoDrop(x, y, actualData)
        return default
    end
end


class TestPane < Wx::Panel

    def initialize(parent)
        super(parent)
#       Wx::Panel.initialize(parent, size=(200, 200))
        evt_paint() {|event| OnPaint(event)}
#       self.Bind(Wx::EVT_PAINT, self.OnPaint)
    end

    def OnPaint(event)
        dc = Wx::PaintDC(self)
        rect = GetClientRect()
        dc.SetPen(Wx::Pen('black'))
        dc.DrawRectangleRect(rect)
        dc.DrawLabel("#{self.class} (#{rect.width}, #{rect.height})",
                     rect.Get(), Wx::ALIGN_CENTER)
    end
end


class NewKeyPane < Wx::Panel

    attr_reader :cancelButton, :okButton

    def initialize(parent, keyHandle=nil)
        super

        @okButton = Wx::Button.new(self, Wx::ID_OK)
        @cancelButton = Wx::Button.new(self, Wx::ID_CANCEL)

        sourceKey = nil
        if keyHandle != nil
            sourceKey = Model.sourcekeys[keyHandle]
        else
            sourceKey = SourceKey.new(nil)
        end
#       assert(sourceKey != nil)

        dispPane  = DispPane.new(self, sourceKey)
        macroPane = MacroPane.new(self, sourceKey)
        modPane   = ModifierKeyPane.new(self)

        leftSizer = Wx::BoxSizer.new(Wx::VERTICAL)
        leftSizer.add(dispPane, 1, Wx::EXPAND)
        leftSizer.add(modPane,  1, Wx::EXPAND)

        upperSizer = Wx::BoxSizer.new(Wx::HORIZONTAL)
        upperSizer.add(leftSizer, 1, Wx::EXPAND)
        upperSizer.add(macroPane, 2, Wx::EXPAND)

        btnSizer = Wx::StdDialogButtonSizer.new()
        btnSizer.affirmative_button = @okButton
        btnSizer.cancel_button = @cancelButton
        btnSizer.realize()

        mainSizer = Wx::BoxSizer.new(Wx::VERTICAL)
        mainSizer.add(upperSizer, 1, Wx::EXPAND)
        mainSizer.add(btnSizer, 0, Wx::EXPAND|Wx::ALL, 10)
        set_sizer(mainSizer)

        mainSizer.fit(self)
        mainSizer.set_size_hints(self)
    end
end


class DispPane < KeyDisplayPane

    def initialize(parent, sourceKey)
        super(parent, $DISPLAY_MODE_DYNAMIC_RESIZE, 10)
        @sourceKey = sourceKey
        @dispKey = DisplayKey.new(sourceKey.handle, Rect.new(0, 0, 1.8, 1.8))
        @displayKeys = {sourceKey.handle => @dispKey}
        _updateKeys()
        _initKeys()
    end

    def _getContentSize
        return Size.new(1.8, 1.8)
    end

#   def OnPaint(self, event)
#       dc = Wx::PaintDC(self)
#       rect = self.GetClientRect()
#       dc.DrawRectangleRect(rect)
#       KeyDisplayPane.OnPaint(self, event)
end

class MacroPane < KeyDisplayPane

    def initialize(parent, sourceKey)
        super(parent, $DISPLAY_MODE_WRAPPED, 10)
        @displayKeys = {}
        _initKeys()
        evt_paint() {|event| OnPaint(event)}
        @keySize = Size.new(6.0, 2.5)
        @keySpacing = Size.new(0.2, 0.2)
        keyRect = Rect.new(0, 0, *@keySize)
        handle = nil
        if sourceKey.usage
            handle = sourceKey.usage.MakeHandle()
        elsif sourceKey.mode
            handle = "mode:%s" % sourceKey.mode
        elsif len(sourceKey.macro) > 0
            for (i, macroKey) in enumerate(sourceKey.macro)
                handle = macroKey.MakeHandle(i)
                @displayKeys[handle] = DisplayKey.new(handle, keyRect)
                @orderedKeyHandles.push(handle)
            end
            handle = nil
        end
        if handle != nil
            dispKey = DisplayKey.new(handle, keyRect)
            @displayKeys[handle] = dispKey
            @orderedKeyHandles.push(handle)
            _updateKeys()
        end
    end

    def _getLabels(displayKey)
        labels = {}
        print displayKey.handle
        keyInfo = displayKey.handle.split(':')
        if keyInfo[0] == 'usage'
            labels['TopLeft'] = "page: #{keyInfo[1]} id: #{keyInfo[2]}"
        elsif keyInfo[0] == 'mode'
            labels['TopLeft'] ="mode: #{keyInfo[1]}"
        elsif keyInfo[0] == 'macro'
            labels['BottomLeft'] = "modifiers: #{keyInfo[4]} "
            labels['TopLeft'] = "page: #{keyInfo[2]} id: #{keyInfo[3]}"
        elsif keyInfo[0] == 'm_usage'
            labels['TopLeft'] = "page: #{keyInfo[2]} id: #{keyInfo[3]}"
        else
            labels['BottomLeft'] = "?"
        end
        return labels
    end

    def OnPaint(event)
        paint do |dc|
            rect = get_client_rect()
            dc.draw_rectangle(rect.x.to_i, rect.y.to_i, rect.width.to_i, rect.height.to_i)
            super
        end
    end
end


class ModifierKeyPane < KeyDisplayPane

    def initialize(parent)
        super(parent, $DISPLAY_MODE_DYNAMIC_RESIZE, 0)
        _initKeys()
    end

    def _initKeys
        super
        @displayKeys = { 'left_shift'=>  DisplayKey.new('left_shift',  Rect.new(0.0, 0.0, 2.8, 1.8)),
                         'left_ctrl'=>   DisplayKey.new('left_ctrl',   Rect.new(0.0, 2.0, 2.8, 1.8)),
                         'left_alt'=>    DisplayKey.new('left_alt',    Rect.new(0.0, 4.0, 2.8, 1.8)),
                         'left_gui'=>    DisplayKey.new('left_gui',    Rect.new(0.0, 6.0, 2.8, 1.8)),
                         'right_shift'=> DisplayKey.new('right_shift', Rect.new(3.0, 0.0, 2.8, 1.8)),
                         'right_ctrl'=>  DisplayKey.new('right_ctrl',  Rect.new(3.0, 2.0, 2.8, 1.8)),
                         'right_alt'=>   DisplayKey.new('right_alt',   Rect.new(3.0, 4.0, 2.8, 1.8)),
                         'right_gui'=>   DisplayKey.new('right_gui',   Rect.new(3.0, 6.0, 2.8, 1.8)) }
        _updateKeys()
    end

    def _getContentSize
        return Size.new(6.0, 8.0)
    end
end

if $0 == __FILE__
    Wx::App.run do
        Model.StaticInit()
        frame = Wx::Frame.new(nil, -1, 'newkey')
        pane = NewKeyPane.new(frame)
        frame.show()
    end
end

